runtime/src/iree/hal/drivers/hip/native_executable.c

```diff
@@ -255,136 +265,181 @@ iree_status_t iree_hal_hip_native_executable_create(
-  IREE_TRACE(uint8_t* export_info_ptr =
-                 ((uint8_t*)executable->modules +
-                  module_count * sizeof(executable->modules[0])));
+  executable->num_devices = topology->count;
+
+  for (iree_host_size_t i = 0; i < topology->count; ++i) {
+    const iree_host_size_t per_device_data_size =
+        topology->count * sizeof(executable->per_device_data[0]);
+    const iree_host_size_t native_executable_device_info_size_offset =
+        (i * native_executable_device_info_size);
+    const uint8_t* per_device_data_location =
+        (uint8_t*)executable + sizeof(*executable);
+
+    executable->per_device_data[i] =
+        (iree_hal_hip_native_executable_per_device_data_t*)(per_device_data_location +
+                                                            per_device_data_size +
+                                                            native_executable_device_info_size_offset);
+  }
 
   // Publish any embedded source files to the tracing infrastructure.
   iree_hal_debug_publish_source_files(
       iree_hal_hip_ExecutableDef_source_files_get(executable_def));
 
-  // Load each module first so that exports can reference them.
   iree_status_t status = iree_ok_status();
-  for (iree_host_size_t i = 0; i < module_count; ++i) {
-    iree_hal_hip_ModuleDef_table_t module_def =
-        iree_hal_hip_ModuleDef_vec_at(modules_vec, i);
-
-    // WARNING: HIP doesn't take an expected length here so we can't bound it.
-    // It's likely that users could craft inputs that read beyond the extents of
-    // the embedded binary.
-    flatbuffers_string_t hsaco_image =
-        iree_hal_hip_ModuleDef_hsaco_image_get(module_def);
-
-    // TODO: pass hipJitOption values to get log info and other info back.
-    // We pass the error buffer today but could use the info log to diagnose
-    // performance warnings.
-    char error_log[8192] = {0};
-    hipJitOption jit_options[] = {
-        hipJitOptionErrorLogBuffer,
-        hipJitOptionErrorLogBufferSizeBytes,
-    };
-    void* jit_option_values[] = {
-        (void*)error_log,
-        (void*)(uint32_t)sizeof(error_log),
-    };
-    hipModule_t module = NULL;
+  for (iree_host_size_t j = 0; j < topology->count && iree_status_is_ok(status);
+       ++j) {
     status = IREE_HIP_RESULT_TO_STATUS(
-        symbols,
-        hipModuleLoadDataEx(&module, hsaco_image, IREE_ARRAYSIZE(jit_options),
-                            jit_options, jit_option_values),
-        "hipModuleLoadDataEx");
-    if (!iree_status_is_ok(status)) {
-      status = iree_status_annotate(
-          status,
-          IREE_SV("mismatched target chip? missing/wrong bitcode directory?"));
-      if (strlen(error_log) > 0) {
-        status =
-            iree_status_annotate(status, iree_make_cstring_view(error_log));
-      }
+        symbols, hipCtxPushCurrent(topology->devices[j].hip_context),
+        "hipCtxPushCurrent");
+
+    if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
       break;
     }
```

Comment:
iree_status_is_ok includes the hot/cold annotation so it can be skipped at use sites
```suggestion
    if (!iree_status_is_ok(status)) break;
```